storage/types.go

```diff
@@ -33,19 +33,61 @@ type Node struct {
 }
 
 // NodeID uniquely identifies a Node within a versioned MerkleTree.
+// NodeIDs as presented to the storage layer will always have a multiple
+// of 8 bits as they identify the root of a 256 element subtree.
+//
+// Reading paths right to left is the natural order when traversing from
+// leaves towards the root. However, for internal nodes the rightmost bits
+// of the IDs are not aligned on a byte boundary so care must be taken.
+//
+// Note that some of the APIs count bits with 0 being the rightmost
+// in the entire path array when some of these might not be significant.
 type NodeID struct {
-	// path is effectively a BigEndian bit set, with path[0] being the MSB
+	// Path is effectively a BigEndian bit set, with Path[0] being the MSB
 	// (identifying the root child), and successive bits identifying the lower
 	// level children down to the leaf.
 	Path []byte
 	// PrefixLenBits is the number of MSB in Path which are considered part of
 	// this NodeID.
 	//
 	// e.g. if Path contains two bytes, and PrefixLenBits is 9, then the 8 bits
-	// in Path[0] are included, along with the lowest bit of Path[1]
+	// in Path[0] are included, along with the MSB of Path[1]. The remaining
+	// 7 bits are not significant.
 	PrefixLenBits int
 }
 
+// Some Additional NodeID examples and Documentation.
+//
+// Consider a hypothetical case of a tree of size 256 and max path length 8.
+// The IDs within this tree will at most have 8 significant bits. To specify
+// the leaf with index 95 requires 8 bits to match the tree height so the
+// Path array would be: [0x5f] with PrefixLength 8. Moving up the tree two
+// levels from this node the depth would be 2 and index 23. This gives a Path
+// array of [0x5c] and PrefixLength is 6, which means the lowest 2 bits are not
+// significant. In binary 0x5c is: 0b01011100 and 23 decimal is: 0b000010111
+// so the index has been shifted left by 2 places in the NodeID. Reading
+// from left to right after 6 bits the node has been reached. The remaining
+// two bits therefore don't matter.
+//
+// Larger tree sizes as used by Logs and Maps work in exactly the same
+// way but with the Path taking up multiple bytes. For example if we move
+// to a tree size of up to 65535 there will be two Path bytes. For a leaf
+// in this tree at depth = 0, index = 19 the resulting NodeID has a
+// Path array of [0x00, 0x13] and PrefixLength 16. To see that it's the
+// left bits that are significant consider depth = 2, index = 16383 in this
+// tree (the rightmost node in a complete tree at this level). The resulting
+// NodeID has Path [0xff, 0xfc] and PrefixLength is 14. When processing this
+// NodeID to move up the tree the lowest 2 bits of Path[1] should be ignored as
+// it only takes 14 bits to reach the root of the tree from level 2 in a 16
+// levels deep tree. Also, testing Bit(0) on this NodeID would not be valid as
```

Comment:
I think that this Bit(0) case is really confusing. I don't see a reason why NodeID must have this multiple-of-8 peculiarity, it's just a bitmask. The code that uses NodeID - yes, it might be easier to implement with such a restriction, but the bitmask code would be clearer without it and still compatible with the rest of the code.

Add a TODO to fix it?

Reply:
I don't think it's that confusing now but maybe I don't understand what you mean. Bit(0) is always well defined, it may just not be a significant bit in a particular NodeID. 

If Go had a suitable type (c.f. Java BitSet) we might have used that. However the patterns must be kept stable as they're used as keys. They would still have to be converted to bytes in some manner as databases don't support bits as a type.

Otherwise, I suggest you write up what you mean as a PR and we'll discuss further.

Reply:
I mean that for a bitmask of 14 bits I would expect `Bit(0)` to return the LSB of the 14, rather than something undefined. In the example above, to get the LSB I need to call `Bit(2)`, which is obv. not a great API.

If any of these IDs are put into storage then it's probably hard to not break compatibility. I still have some hope that we use it only as keys of in-memory data structures, as then it would be possible to change bits layout. Not in this PR, you are right.

Reply:
As I said above they are used in storage as keys. The serialized format can't be changed without major pain.

But I think that the bitwise access functions are actually only used by the sparse Merkle tree code and some tests. In the sparse tree case all the paths are fixed height in multiples of 8 so Bit(0) is always valid.